File: 字母大小写全排列.cpp
class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> ans;
        int n = s.size();

        function<void(int)> dfs = [&](int i) -> void {
            if (i == n) {
                ans.emplace_back(s);
                return;
            }

            if (isalpha(s[i])) {
                s[i] = tolower(s[i]);
                dfs(i + 1);
                s[i] = toupper(s[i]);
                dfs(i + 1);
            } else {
                dfs(i + 1);
            }
        };
        dfs(0);
        return ans;
    }
};

File: 字母大小写全排列.py
class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        ans = []
        s = list(s)
        n = len(s)

        def dfs(i):
            if i == n:
                ans.append("".join(s))
                return
            if s[i].isalpha():
                s[i] = s[i].lower()
                dfs(i + 1)
                s[i] = s[i].upper()
                dfs(i + 1)
            else:
                dfs(i + 1)
        
        dfs(0)
        return ans
